plugins/inputs/nsdp/nsdp.go

```diff
@@ -0,0 +1,121 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package nsdp
+
+import (
+	_ "embed"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/tdrn-org/go-nsdp"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+const pluginName = "nsdp"
+
+const defaultTimeout = config.Duration(2 * time.Second)
+
+type NSDP struct {
+	Target      string          `toml:"target"`
+	DeviceLimit uint            `toml:"device_limit"`
+	Timeout     config.Duration `toml:"timeout"`
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func defaultNSDP() *NSDP {
+	return &NSDP{
+		Target:  nsdp.IPv4BroadcastTarget,
+		Timeout: defaultTimeout,
+	}
+}
+
+func (*NSDP) SampleConfig() string {
+	return sampleConfig
+}
+
+func (plugin *NSDP) Init() error {
+	return nil
+}
+
+func (plugin *NSDP) Gather(acc telegraf.Accumulator) error {
+	plugin.Log.Debugf("Querying %s", plugin.Target)
+	conn, request, err := plugin.prepareGatherRequest()
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	responses, err := conn.SendReceiveMessage(request)
+	if err != nil {
+		return err
+	}
+	for device, response := range responses {
+		plugin.Log.Debugf("Processing device: %s", device)
+		plugin.gatherDevice(acc, device, response)
+	}
+	return nil
+}
+
+func (plugin *NSDP) prepareGatherRequest() (*nsdp.Conn, *nsdp.Message, error) {
+	conn, err := nsdp.NewConn(plugin.Target, plugin.Log.Level().Includes(telegraf.Debug))
+	if err != nil {
+		return nil, nil, err
+	}
+	conn.ReceiveDeviceLimit = plugin.DeviceLimit
+	conn.ReceiveTimeout = time.Duration(plugin.Timeout)
+	request := nsdp.NewMessage(nsdp.ReadRequest)
+	request.AppendTLV(nsdp.EmptyDeviceModel())
+	request.AppendTLV(nsdp.EmptyDeviceName())
+	request.AppendTLV(nsdp.EmptyDeviceIP())
+	request.AppendTLV(nsdp.EmptyPortStatistic())
+	return conn, request, nil
+}
+
+func (plugin *NSDP) gatherDevice(acc telegraf.Accumulator, device string, response *nsdp.Message) {
+	var deviceModel string
+	var deviceName string
+	var deviceIP net.IP
+	portStatistics := make(map[uint8]*nsdp.PortStatistic, 0)
+	for _, tlv := range response.Body {
+		switch tlv.Type() {
+		case nsdp.TypeDeviceModel:
+			deviceModel = tlv.(*nsdp.DeviceModel).Model
+		case nsdp.TypeDeviceName:
+			deviceName = tlv.(*nsdp.DeviceName).Name
+		case nsdp.TypeDeviceIP:
+			deviceIP = tlv.(*nsdp.DeviceIP).IP
+		case nsdp.TypePortStatistic:
+			portStatistic := tlv.(*nsdp.PortStatistic)
+			portStatistics[portStatistic.Port] = portStatistic
+		}
+	}
+	for port, statistic := range portStatistics {
+		if statistic.Received != 0 || statistic.Sent != 0 {
```

Comment:
Why do you omit inactive ports? This is kind of unexpected...

Reply:
Intention was to filter out inactive ports to avoid excessive lines of zero-stats. I have removed the filter as part of 8c321f127f426a5c1cbdc9876b56cf4409fafce2